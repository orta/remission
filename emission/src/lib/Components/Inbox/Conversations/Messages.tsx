import * as React from "react"
import { ActivityIndicator, Dimensions, FlatList, RefreshControl } from "react-native"
import { createPaginationContainer, graphql, RelayPaginationProp } from "react-relay"

import ARSwitchBoard from "../../../NativeModules/SwitchBoard"
import Message from "./Message"
import ArtworkPreview from "./Preview/ArtworkPreview"
import ShowPreview from "./Preview/ShowPreview"

interface Props {
  conversation?: RelayProps["me"]["conversation"]
  relay?: RelayPaginationProp
  onDataFetching?: (loading: boolean) => void
}

interface State {
  fetchingMoreData: boolean
  reloadingData: boolean
  shouldStickFirstMessageToTop: boolean
}

export class Messages extends React.Component<Props, State> {
  constructor(props) {
    super(props)

    this.state = {
      fetchingMoreData: false,
      reloadingData: false,
      shouldStickFirstMessageToTop: false,
    }
  }

  renderMessage({ item, index }) {
    const conversationItem = this.props.conversation.items[0].item
    const conversation = this.props.conversation
    const partnerName = conversation.to.name
    const senderName = item.is_from_user ? conversation.from.name : partnerName
    const initials = item.is_from_user ? conversation.from.initials : conversation.to.initials
    return (
      <Message
        index={index}
        firstMessage={item.first_message}
        initialText={conversation.initial_message}
        message={item}
        senderName={senderName}
        initials={initials}
        artworkPreview={
          item.first_message &&
          conversationItem.__typename === "Artwork" &&
          <ArtworkPreview
            artwork={conversationItem}
            onSelected={() => ARSwitchBoard.presentNavigationViewController(this, conversationItem.href)}
          />
        }
        showPreview={
          item.first_message &&
          conversationItem.__typename === "Show" &&
          <ShowPreview
            show={conversationItem}
            onSelected={() => ARSwitchBoard.presentNavigationViewController(this, conversationItem.href)}
          />
        }
      />
    )
  }

  loadMore() {
    if (!this.props.relay.hasMore() || this.props.relay.isLoading()) {
      return
    }

    const updateState = (loading: boolean) => {
      this.setState({ fetchingMoreData: loading })
      if (this.props.onDataFetching) {
        this.props.onDataFetching(loading)
      }
    }

    updateState(true)
    this.props.relay.loadMore(10, e => {
      updateState(false)
    })
  }

  reload() {
    const count = this.props.conversation.messages.edges.length
    this.setState({ reloadingData: true })
    this.props.relay.refetchConnection(count, e => {
      this.setState({ reloadingData: false })
    })
  }

  render() {
    const edges = (this.props.conversation.messages || { edges: [] }).edges
    const messageCount = edges.length
    const messages = edges.map((edge, index) => {
      const isFirstMessage = this.props.relay && !this.props.relay.hasMore() && index === messageCount - 1
      return { first_message: isFirstMessage, key: edge.cursor, ...edge.node }
    })
    const refreshControl = <RefreshControl refreshing={this.state.reloadingData} onRefresh={this.reload.bind(this)} />

    return (
      <FlatList
        inverted={!this.state.shouldStickFirstMessageToTop}
        data={this.state.shouldStickFirstMessageToTop ? messages.reverse() : messages}
        renderItem={this.renderMessage.bind(this)}
        onEndReached={this.loadMore.bind(this)}
        onEndReachedThreshold={0.2}
        onContentSizeChange={(width, height) => {
          // If there aren't enough items to scroll through
          // display messages from the top
          const windowHeight = Dimensions.get("window").height
          const containerHeight = windowHeight - 100

          this.setState({
            shouldStickFirstMessageToTop: height < containerHeight,
          })
        }}
        refreshControl={refreshControl}
        ListFooterComponent={<ActivityIndicator animating={this.state.fetchingMoreData} hidesWhenStopped />}
      />
    )
  }
}

export default createPaginationContainer(
  Messages,
  {
    conversation: graphql.experimental`
      fragment Messages_conversation on Conversation
        @argumentDefinitions(count: { type: "Int", defaultValue: 10 }, after: { type: "String" }) {
        __id
        id
        from {
          name
          email
          initials
        }
        to {
          name
          initials
        }
        initial_message
        messages(first: $count, after: $after, sort: DESC) @connection(key: "Messages_messages", filters: []) {
          pageInfo {
            startCursor
            endCursor
            hasPreviousPage
            hasNextPage
          }
          edges {
            cursor
            node {
              impulse_id
              is_from_user
              body
              ...Message_message
            }
          }
        }
        items {
          item {
            ... on Artwork {
              __typename
              href
              ...ArtworkPreview_artwork
            }
            ... on Show {
              __typename
              href
              ...ShowPreview_show
            }
          }
        }
      }
    `,
  },
  {
    direction: "forward",
    getConnectionFromProps(props) {
      return props.conversation && props.conversation.messages
    },
    getFragmentVariables(prevVars, totalCount) {
      return {
        ...prevVars,
        count: totalCount,
      }
    },
    getVariables(props, paginationInfo, fragmentVariables) {
      return {
        conversationID: props.conversation.id,
        count: paginationInfo.count,
        after: paginationInfo.cursor,
      }
    },
    query: graphql.experimental`
      query MessagesQuery($conversationID: String!, $count: Int!, $after: String) {
        me {
          conversation(id: $conversationID) {
            ...Messages_conversation @arguments(count: $count, after: $after)
          }
        }
      }
    `,
  }
)

interface RelayProps {
  me: {
    conversation: {
      id: string
      __id: string
      to: {
        name: string
        initials: string
      }
      from: {
        name: string
        email: string
        initials: string
      }
      initial_message: string
      messages: {
        pageInfo?: {
          hasNextPage: boolean
        }
        edges: Array<{
          cursor: string
          node: {
            impulse_id: string
            is_from_user: boolean
          } | null
        }>
      }
      items: Array<{
        item: any
      }>
    }
  }
}
