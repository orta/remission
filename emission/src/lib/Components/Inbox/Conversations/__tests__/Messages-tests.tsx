import * as React from "react"

import "react-native"
import * as renderer from "react-test-renderer"
import Messages from "../Messages"

it("looks correct when rendered", () => {
  const messages = renderer.create(<Messages conversation={props} />) as any
  expect(messages).toMatchSnapshot()
})

const props = {
  id: "420",
  from: {
    name: "Anita Garibaldi",
    email: "anita@garibaldi.br",
    initials: "AG",
  },
  to: { name: "Kimberly Klark", initials: "KK" },
  initial_message: "Adoro! Por favor envie-me mais informações",
  messages: {
    edges: [
      {
        node: {
          id: 222,
          body: "Adoro! Por favor envie-me mais informações",
          from_email_address: "anita@garibaldi.br",
          attachments: [],
          from: {
            name: "Percy",
            email: "percy@cat.com",
          },
        },
      },
    ],
  },
  items: [
    {
      title: "The Mythic Being: Sol’s Drawing #1–5",
      item: {
        __typename: "Artwork",
        id: "adrian-piper-the-mythic-being-sols-drawing-number-1-5",
        href: "/artwork/adrian-piper-the-mythic-being-sols-drawing-number-1-5",
        title: "The Mythic Being: Sol’s Drawing #1–5",
        date: "1974",
        artist_names: "Adrian Piper",
        image: {
          url: "https://d32dm0rphc51dk.cloudfront.net/W1FkNoM9IjrND_xv_DTkeg/normalized.jpg",
          image_url: "https://d32dm0rphc51dk.cloudfront.net/J0uofgV9e8cIxGiZwn12mg/:version.jpg",
        },
      },
    },
  ],
}
