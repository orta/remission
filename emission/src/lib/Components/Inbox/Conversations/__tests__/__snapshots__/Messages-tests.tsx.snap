// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`looks correct when rendered 1`] = `
<RCTScrollView
  ListFooterComponent={
    <ActivityIndicator
      animating={false}
      color="#999999"
      hidesWhenStopped={true}
      size="small"
    />
  }
  data={
    Array [
      Object {
        "attachments": Array [],
        "body": "Adoro! Por favor envie-me mais informações",
        "first_message": undefined,
        "from": Object {
          "email": "percy@cat.com",
          "name": "Percy",
        },
        "from_email_address": "anita@garibaldi.br",
        "id": 222,
        "key": undefined,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  inverted={true}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onViewableItemsChanged={undefined}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "transform": Array [
          Object {
            "scaleY": -1,
          },
        ],
      },
      undefined,
    ]
  }
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingLeft": 20,
              "paddingRight": 20,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Array [
                Object {
                  "alignSelf": "stretch",
                  "marginBottom": 10,
                  "marginTop": 15,
                },
                undefined,
              ],
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#e5e5e5",
                  "borderRadius": 15,
                  "height": 30,
                  "width": 30,
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "alignSelf": "center",
                    "color": "#666666",
                    "fontFamily": "Avant Garde Gothic ITCW01Dm",
                    "fontSize": 10,
                    "marginTop": 8,
                  },
                  undefined,
                ]
              }
              user={undefined}
            >
              KK
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "flexBasis": 0,
                  "flexDirection": "column",
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
                Array [
                  Object {
                    "marginLeft": 10,
                  },
                  undefined,
                ],
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Array [
                    Object {
                      "alignSelf": "stretch",
                      "marginBottom": 10,
                    },
                    undefined,
                  ],
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "black",
                      "fontSize": 11,
                      "textAlign": "left",
                    },
                    Object {
                      "color": "#cccccc",
                    },
                    Array [
                      Object {
                        "fontSize": 11.5,
                        "marginRight": 6,
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": "Avant Garde Gothic ITCW01Dm",
                    },
                  ]
                }
              >
                KIMBERLY KLARK
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "#cccccc",
                      "fontSize": 11,
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "color": "#f1af1b",
                        "fontSize": 11.5,
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": "Avant Garde Gothic ITCW01Dm",
                    },
                  ]
                }
              >
                PENDING
              </Text>
            </View>
            <View
              linkStyle={
                Object {
                  "color": "#0645ad",
                }
              }
              onPress={[Function]}
              style={undefined}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "black",
                      "fontSize": 16,
                      "textAlign": "left",
                    },
                    Object {
                      "color": "#cccccc",
                    },
                    Object {},
                    Object {
                      "fontFamily": "AGaramondPro-Regular",
                    },
                  ]
                }
              >
                Adoro! Por favor envie-me mais informações
              </Text>
            </View>
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#cccccc",
                    "fontFamily": "AGaramondPro-Regular",
                  },
                  Array [
                    Object {
                      "marginTop": 10,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Percy · percy@cat.com
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        }
      }
    >
      <ActivityIndicator
        animating={false}
        color="#999999"
        hidesWhenStopped={true}
        size="small"
      />
    </View>
  </View>
</RCTScrollView>
`;
